src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -644,26 +646,41 @@ private void Dispose()
                 {
                     if (!_noImplicitTransforms)
                     {
+                        // given the type, fetch the implicit transforms for that type and put it in the implicitTransforms variable.  
                         Type argType = args.GetType();
-                        if (_expectedArgType != argType)
+                        TransformSpec implicitTransforms;
+
+                        // First check the one-element cache _firstImplicitTransformsEntry
+                        ImplicitTransformEntry cacheEntry = _firstImplicitTransformsEntry;
+                        if (cacheEntry != null && cacheEntry.Type == argType)
+                        {
+                            implicitTransforms = cacheEntry.Transforms;     // Yeah we hit the cache.  
+                        }
+                        else if (cacheEntry == null)
                         {
-                            // Figure out the default properties to send on to EventSource.  These are all string or primitive properties.  
-                            _implicitTransforms = null;
-                            TransformSpec newSerializableArgs = null;
-                            TypeInfo curTypeInfo = argType.GetTypeInfo();
-                            foreach (var property in curTypeInfo.DeclaredProperties)
+                            // _firstImplicitTransformsEntry, we should fill it.  
+                            implicitTransforms = MakeImplicitTransforms(argType);
+                            Interlocked.CompareExchange(ref _firstImplicitTransformsEntry, 
+                                new ImplicitTransformEntry() { Type = argType, Transforms = implicitTransforms }, null);
+                        }
+                        else
+                        {
+                            // This should only happen when you are wildcarding your events (reasonably rare).   
+                            // In that case you will probably need many types
+                            // Note currently we don't limit the cache size, but it is limited by the number of 
+                            // distinct types of objects passed to DiagnosticSource.Write.  
+                            if (_implicitTransformsTable == null)
                             {
-                                var propertyType = property.PropertyType;
-                                newSerializableArgs = new TransformSpec(property.Name, 0, property.Name.Length, newSerializableArgs);
+                                Interlocked.CompareExchange(ref _implicitTransformsTable,
+                                    new ConcurrentDictionary<Type, TransformSpec>(1, 8), null);
                             }
-                            _expectedArgType = argType;
-                            _implicitTransforms = Reverse(newSerializableArgs);
+                            implicitTransforms = _implicitTransformsTable.GetOrAdd(argType, MakeImplicitTransforms);
```

Comment:
This is going to allocate a delegate for MakeImplicitTransforms on every call.  Is that ok perf wise?  Unfortunately the C# compiler doesn't currently cache delegates for method groups.  It does for lambdas, though, so since MakeImplicitTransforms is static, if you were to change this to instead be:
```C#
implicitTransforms = _implicitTransformsTable.GetOrAdd(argType, k => MakeImplicitTransforms(k));
```
then it would cache the delegate.

Reply:
I did not know that C# was not caching my static delegate, I have changed it to what you suggest.

Reply:
Yeah, unfortunately.  Still an open issue: https://github.com/dotnet/roslyn/issues/5835